{
 "cells": [
  {
   "cell_type": "markdown",
   "source": [
    "# Footprint calling and plotting of a single DHS\n",
    "\n",
    "This notebook shows an example of how call footprints on a single region and plot the results. This code recreates **Extended Data Fig. 1** from Vierstra *et al.* 2020.\n",
    "\n",
    "The plotting performed in this example relies on the python package `genome-tools`, which provides an assortment of ways to plot genomic data. You can find a short tutorial in the form of a python notebook [here](https://github.com/jvierstra/genome-tools/blob/master/examples/UsageExamples.ipynb) on the various features."
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scipy as sc\n",
    "import scipy.stats as stats\n",
    "\n",
    "import pysam\n",
    "\n",
    "from genome_tools import bed, genomic_interval\n",
    "from footprint_tools import cutcounts\n",
    "from footprint_tools.modeling import bias, predict, dispersion\n",
    "from footprint_tools.stats import fdr, windowing, utils\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.gridspec as gridspec\n",
    "from matplotlib.ticker import MaxNLocator"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "source": [
    "# We can grab the FASTA, BAM and dispersion model files remotely with the API\n",
    "fasta_file = \"https://resources.altius.org/~jvierstra/projects/footprinting.2020/hg38.all.fa\"\n",
    "\n",
    "base_url=\"https://resources.altius.org/~jvierstra/projects/footprinting.2020/per.dataset/CD20+-DS18208/\"\n",
    "bam_file=base_url + \"reads.bam\"\n",
    "disp_model_file=base_url + \"dm.json\"\n",
    "\n",
    "!curl https://resources.altius.org/~jvierstra/projects/footprinting.2020/vierstra_et_al.6mer-model.txt -o bm.txt\n",
    "bias_model_file = \"bm.txt\" \n",
    "\n",
    "# This is the genomic region to plot data\n",
    "interval = genomic_interval('chr19', 48363826, 48364602)"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [
    "## Perform footprint detection"
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "source": [
    "counts_reader = cutcounts.bamfile(bam_file)\n",
    "fasta_reader = pysam.FastaFile(fasta_file)\n",
    "dm = dispersion.load_dispersion_model(disp_model_file)\n",
    "bm = bias.kmer_model(bias_model_file)\n",
    "\n",
    "predictor = predict.prediction(counts_reader, fasta_reader, bm, half_win_width = 5, smoothing_half_win_width = 50, smoothing_clip = 0.01)\n",
    "(obs_counts, exp_counts, win_counts) = predictor.compute(interval)\n",
    "    \n",
    "# Compute expected counts\n",
    "# Note: windowed counts are used to generated expected counts\n",
    "# and not used directly in statistical testing\n",
    "obs = obs_counts['+'][1:] + obs_counts['-'][:-1]\n",
    "exp = exp_counts['+'][1:] + exp_counts['-'][:-1]\n",
    "win = win_counts['+'][1:] + win_counts['-'][:-1]\n",
    "\n",
    "# Per-nucleotide p-values\n",
    "pvals = dm.p_values(exp, obs)\n",
    "\n",
    "# Windowed p-values with Stouffer's Z\n",
    "winpvals_func = lambda x: windowing.stouffers_z(np.ascontiguousarray(x), 3)\n",
    "winpvals = np.array(winpvals_func(pvals))\n",
    "\n",
    "# Resample from expected distributions for emperical adjustment of p-values for multiple testing\n",
    "_, pvals_null = dm.sample(exp, 1000)\n",
    "winpvals_null = np.apply_along_axis(winpvals_func, 0, pvals_null)\n",
    "fpr = fdr.emperical_fdr(winpvals_null, winpvals)\n",
    "\n",
    "# Call footprints at 0.01 FPR\n",
    "fps = [genomic_interval(interval.chrom, interval.start+x, interval.start+y) \n",
    "    for x, y in utils.segment(np.array(fpr), 0.01, 3, decreasing=True) ]\n"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [
    "## Plot genomic footprinting data"
   ],
   "metadata": {}
  },
  {
   "cell_type": "raw",
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.gridspec as gridspec\n",
    "from matplotlib.ticker import MaxNLocator\n",
    "from matplotlib.collections import BrokenBarHCollection\n",
    "\n",
    "def discrete_cmap(N, base_cmap=None):\n",
    "    \"\"\"Create an N-bin discrete colormap from the specified input map\"\"\"\n",
    "\n",
    "    # Note that if base_cmap is a string or None, you can simply do\n",
    "    #    return plt.cm.get_cmap(base_cmap, N)\n",
    "    # The following works for string, None, or a colormap instance:\n",
    "\n",
    "    base = plt.cm.get_cmap(base_cmap)\n",
    "    color_list = base(np.linspace(0, 1, N))\n",
    "    cmap_name = base.name + str(N)\n",
    "    return base.from_list(cmap_name, color_list, N)\n",
    "\n",
    "import seaborn as sns"
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "source": [
    "from genome_tools.plotting import pwm, ideogram\n",
    "from genome_tools.plotting import gencode_annotation_track, continuous_data_track, segment_track\n",
    "from genome_tools.plotting.connectors import zoom_effect\n",
    "\n",
    "def make_nb_plot(e, o, dm, ax, lo = 0, hi = 125):\n",
    "    \"\"\"Plots the dispersion model negative binomial distribution\"\"\"\n",
    "    \n",
    "    xx = np.arange(lo, hi)\n",
    "    \n",
    "    mu = dm.fit_mu(e)\n",
    "    r = dm.fit_r(e)\n",
    "    p = r/(r+mu)\n",
    "\n",
    "    yy = sc.stats.nbinom.pmf(xx, r, p)\n",
    "    ax.plot(xx, yy, label = \"Expected cleavage rate\")\n",
    "    ax.fill_between(xx[:int(o)], 0, yy[:int(o)], edgecolor = 'none', label=\"Prob. observed\")\n",
    "    \n",
    "    ax.set_xlim(left = lo, right = hi)\n",
    "    \n",
    "    ax.set_xlabel(\"Cleavages\")\n",
    "    ax.set_ylabel(\"Density\")\n",
    "\n",
    "    [ax.spines[loc].set_color(\"none\") for loc in [\"top\", \"right\"]]\n",
    "    ax.xaxis.set_ticks_position(\"bottom\")\n",
    "    ax.xaxis.set_tick_params(direction = \"out\")\n",
    "    ax.xaxis.set(major_locator = MaxNLocator(4))\n",
    "    \n",
    "    ax.yaxis.set_ticks_position(\"left\")\n",
    "    ax.yaxis.set_tick_params(direction = \"out\")\n",
    "    ax.yaxis.set(major_locator = MaxNLocator(3))\n",
    "    ax.set_ylim(bottom=0)\n",
    "    "
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "raw",
   "source": [
    "# Positions to highlight with distribution\n",
    "pos = [278, 536]\n",
    "\n",
    "# Footprint thresholds to plot\n",
    "thresholds=[0.05, 0.01, 0.001, 0.0001]"
   ],
   "metadata": {}
  },
  {
   "cell_type": "raw",
   "source": [
    "fig = plt.figure()\n",
    "fig.set_size_inches(6, 10)\n",
    "\n",
    "gs = gridspec.GridSpec(4, 1, height_ratios=[2, 2, 1.5, 3], wspace=0.2, hspace=0.5)\n",
    "\n",
    "# Plot annotations\n",
    "\n",
    "gs_annot = gridspec.GridSpecFromSubplotSpec(1,1, subplot_spec = gs[0,:], hspace = 0)\n",
    "\n",
    "ax_genes = fig.add_subplot(gs_annot[0,:])\n",
    "\n",
    "# Gencode annotation\n",
    "gencode_file_url = \"https://resources.altius.org/~jvierstra/projects/footprinting.2020/annotations/\"\n",
    "gencode_file_url += \"gencode.v25.chr_patch_hapl_scaff.basic.annotation.gff3.gz\"\n",
    "\n",
    "trk_genes = gencode_annotation_track(interval)\n",
    "trk_genes.load_data(gencode_file_url)\n",
    "trk_genes.render(ax_genes)\n",
    "\n",
    "ax_genes.xaxis.set_visible(True)\n",
    "ax_genes.xaxis.set_ticks_position('top')\n",
    "ax_genes.spines['top'].set_color('black')\n",
    "\n",
    "# Plot observed/expected cleavages\n",
    "\n",
    "gs_cleavage = gridspec.GridSpecFromSubplotSpec(2,1, subplot_spec = gs[1,:], hspace = 0.3)\n",
    "\n",
    "cmap=discrete_cmap(4, \"YlGnBu_r\")\n",
    "\n",
    "ax_obs = fig.add_subplot(gs_cleavage[0,:])\n",
    "trk = continuous_data_track(interval, obs, fill_between=0, min=0, max=100, facecolor=cmap(0),edgecolor=cmap(0), lw=0.5, scale_bar=100)\n",
    "trk.render(ax_obs)\n",
    "\n",
    "[ax_obs.axvline(i+interval.start, color='red') for i in pos]\n",
    "ax_obs.set_ylabel(\"Observed\", rotation=0, ha=\"right\")\n",
    "\n",
    "ax_exp = fig.add_subplot(gs_cleavage[1,:])\n",
    "trk = continuous_data_track(interval, exp, fill_between=0, min=0, max=100, facecolor=cmap(1), edgecolor=cmap(1), lw=0.5)\n",
    "trk.render(ax_exp)\n",
    "\n",
    "[ax_exp.axvline(i+interval.start, color='red') for i in pos]\n",
    "ax_exp.set_ylabel(\"Expected\", rotation=0, ha=\"right\")\n",
    "\n",
    "# Plot negative binomials\n",
    "\n",
    "gs_nb = gridspec.GridSpecFromSubplotSpec(1,len(pos), subplot_spec = gs[2,:], hspace = 0, wspace=0.5)\n",
    "\n",
    "for i, j in enumerate(pos):\n",
    "    ax_nb = fig.add_subplot(gs_nb[0,i])\n",
    "    make_nb_plot(exp[j], obs[j], dm, ax_nb, hi=80)\n",
    "    if i==0:\n",
    "        ax_nb.legend(loc=9, bbox_to_anchor=[0.5, 1.5])\n",
    "\n",
    "        \n",
    "# Plot cleavage statistics and footprint calls\n",
    "\n",
    "gs_stats = gridspec.GridSpecFromSubplotSpec(3,1, subplot_spec = gs[3,:], hspace = 0.3)\n",
    "\n",
    "cmap=discrete_cmap(4, \"Reds\")\n",
    "\n",
    "ax_lnp = fig.add_subplot(gs_stats[0,:])\n",
    "trk = continuous_data_track(interval, -np.log10(pvals), fill_between=0, min=0, max=5, facecolor=cmap(2), edgecolor=cmap(2), lw=0.5)\n",
    "trk.render(ax_lnp)\n",
    "\n",
    "[ax_lnp.axvline(i+interval.start, color='k') for i in pos]\n",
    "ax_lnp.set_ylabel(\"-log10 P\", rotation=0, ha=\"right\")\n",
    "\n",
    "\n",
    "ax_winlnp = fig.add_subplot(gs_stats[1,:])\n",
    "trk = continuous_data_track(interval, -np.log10(winpvals), fill_between=0, min=0, max=10, facecolor=cmap(3), edgecolor=cmap(3), lw=0.5)\n",
    "trk.render(ax_winlnp)\n",
    "\n",
    "ax_winlnp.set_ylabel(\"Windowed -log10 P\", rotation=0, ha=\"right\")\n",
    "\n",
    "# Plot footprints\n",
    "ax_fps = fig.add_subplot(gs_stats[2,:])\n",
    "\n",
    "cmap=discrete_cmap(len(thresholds), \"RdPu_r\")\n",
    "for i, thresh in enumerate(sorted(thresholds)):\n",
    "    fps = [genomic_interval(interval.chrom, interval.start+x, interval.start+y) for x, y in segment.segment(-np.array(fpr), -thresh, 3) ]\n",
    "    starts=[fp.start for fp in fps]\n",
    "    w=[len(fp) for fp in fps]\n",
    "\n",
    "    coll = BrokenBarHCollection(zip(starts,w), (i+0.1, 0.8), facecolors=cmap(i), edgecolors='k', linewidths=0.5)\n",
    "    ax_fps.add_collection(coll)\n",
    "    \n",
    "    ax_fps.axhline(i+0.5, color='lightgrey', ls='dashed', lw=0.8, zorder=-1)\n",
    "\n",
    "ax_fps.set_xlim(interval.start, interval.end)\n",
    "ax_fps.set_ylim(0, 4)\n",
    "ax_fps.xaxis.set_visible(False)\n",
    "\n",
    "ax_fps.yaxis.set_ticks(np.arange(len(thresholds))+0.5)\n",
    "ax_fps.yaxis.set_ticklabels(sorted(thresholds))\n",
    "\n",
    "ax_fps.set_ylabel(\"Footprints (FDR)\", rotation=0, ha=\"right\")\n",
    "\n",
    "sns.despine(ax=ax_fps, bottom=True)\n",
    "\n",
    "plt.show()"
   ],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [
    "## Auto-correlation analysis of *p*-values"
   ],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [
    "The following computes the Pearson's correlations between null p-values at offsets up to 25 bp."
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "source": [
    "m, n = pvals_null.shape\n",
    "\n",
    "ts = range(0, 26)\n",
    "\n",
    "cor_mat = np.ones((m, len(ts)), dtype = float)\n",
    "for i in range(m):\n",
    "    for t in ts[1:]:        \n",
    "        x = pvals_null[:-t, i]\n",
    "        y = pvals_null[t:,i]\n",
    "        cor_mat[i, t] = np.corrcoef(x, y)[0, 1]\n",
    "\n",
    "err = np.percentile(cor_mat, [10, 90], axis = 0)"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "source": [
    "fig, ax = plt.subplots()\n",
    "\n",
    "ax.plot(ts[1:], np.mean(cor_mat, axis = 0)[1:], lw = 2, label = \"Mean correlation (r)\")\n",
    "ax.fill_between(ts[1:], err[0,1:], err[1,1:], color = \"lightgrey\", label = \"95% CI\")\n",
    "\n",
    "ax.axhline(0.0, ls = \"--\", color = 'black')\n",
    "\n",
    "ax.set_xlim(0, 26)\n",
    "ax.set_ylim(-0.4, 0.4)\n",
    "\n",
    "ax.set_xlabel(\"p-value offset (nt)\")\n",
    "ax.set_ylabel(\"Correlation (Pearson's r)\")\n",
    "\n",
    "\n",
    "[ax.spines[loc].set_color(\"none\") for loc in [\"top\", \"right\"]]\n",
    "\n",
    "ax.xaxis.set_ticks([1, 5, 10, 15, 20, 25])\n",
    "ax.xaxis.set_ticks_position(\"bottom\")\n",
    "ax.xaxis.set_tick_params(direction = \"out\")\n",
    "\n",
    "ax.yaxis.set_ticks_position(\"left\")\n",
    "ax.yaxis.set_tick_params(direction = \"out\")\n",
    "ax.yaxis.set(major_locator = MaxNLocator(4))\n",
    "    \n",
    "ax.legend()\n",
    "\n",
    "fig.set_size_inches(3, 2)\n",
    "\n",
    "plt.show()"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [
    "## Plot obserbved and null *p*-value distributions"
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "source": [
    "fig = plt.figure()\n",
    "fig.set_size_inches(3, 6)\n",
    "\n",
    "gs = gridspec.GridSpec(3, 1, hspace=0.4)\n",
    "\n",
    "ax = fig.add_subplot(gs[0, 0])\n",
    "ax.hist(winpvals, bins = np.arange(0, 1.05, 0.05), lw = 0, facecolor = \"tomato\", density = True)\n",
    "ax.set_ylim(bottom = 0, top = 7.5)\n",
    "\n",
    "ax.set_xlabel(\"Windowed p-value (observed)\")\n",
    "ax.set_ylabel(\"Density\")\n",
    "\n",
    "[ax.spines[loc].set_color('none') for loc in [\"right\", \"top\"]]\n",
    "ax.xaxis.set(ticks_position = 'bottom')\n",
    "ax.xaxis.set_tick_params(direction = \"out\")\n",
    "\n",
    "ax.yaxis.set(ticks_position = 'left', major_locator = MaxNLocator(4))\n",
    "ax.yaxis.set_tick_params(direction = \"out\")\n",
    "\n",
    "ax.set_ylim(bottom = 0, top = 7.5)\n",
    "\n",
    "ax = fig.add_subplot(gs[1, 0])\n",
    "ax.hist(winpvals_null[:,1], bins = np.arange(0, 1.05, 0.05), lw = 0, facecolor = \"mediumpurple\", density = True)\n",
    "#ax.hist(winpvals_null, bins = np.arange(0, 1.05, 0.05), lw = 0, facecolor = \"green\", normed = True)\n",
    "ax.set_ylim(bottom = 0, top = 7.5)\n",
    "\n",
    "ax.set_xlabel(\"Windowed p-value (resampled null)\")\n",
    "ax.set_ylabel(\"Density\")\n",
    "\n",
    "[ax.spines[loc].set_color('none') for loc in [\"right\", \"top\"]]\n",
    "ax.xaxis.set(ticks_position = 'bottom')\n",
    "ax.xaxis.set_tick_params(direction = \"out\")\n",
    "\n",
    "ax.yaxis.set(ticks_position = 'left', major_locator = MaxNLocator(4))\n",
    "ax.yaxis.set_tick_params(direction = \"out\")\n",
    "\n",
    "ax.margins(x = 0.025)\n",
    "\n",
    "ax = fig.add_subplot(gs[2, 0])\n",
    "ax.plot(np.sort(fpr), lw = 3)\n",
    "\n",
    "[ax.spines[loc].set_color('none') for loc in [\"right\", \"top\"]]\n",
    "ax.xaxis.set(ticks_position = 'bottom')\n",
    "ax.xaxis.set_tick_params(direction = \"out\")\n",
    "\n",
    "ax.yaxis.set(ticks_position = 'left', ticks = [ 0.0, 0.25, 0.5, 0.75, 1.0])\n",
    "ax.yaxis.set_tick_params(direction = \"out\")\n",
    "\n",
    "ax.margins(x = 0.025)\n",
    "\n",
    "ax.set_xlabel(\"Nucleotides sorted by p-value\")\n",
    "ax.set_ylabel(\"False positive rate\")\n",
    "\n",
    "plt.show()"
   ],
   "outputs": [],
   "metadata": {}
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}